pkg/controller/atlasdeployment/backup.go

```diff
@@ -0,0 +1,287 @@
+package atlasdeployment
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1/status"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/customresource"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/watch"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/workflow"
+
+	"go.mongodb.org/atlas/mongodbatlas"
+	"golang.org/x/sync/errgroup"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	mdbv1 "github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1"
+)
+
+func (r *AtlasDeploymentReconciler) ensureBackupScheduleAndPolicy(
+	ctx context.Context,
+	service *workflow.Context,
+	projectID string,
+	clusterName string,
+	backupScheduleRef types.NamespacedName,
+	enabled bool,
+	requestNamespacedName client.ObjectKey,
+) error {
+	if backupScheduleRef.Name == "" || backupScheduleRef.Namespace == "" {
+		r.Log.Debug("no backup schedule configured for the deployment")
+
+		err := r.garbageCollectBackupResource(ctx, clusterName)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	if !enabled {
+		return fmt.Errorf("can not proceed with backup configuration. Backups are not enabled for cluster %s", clusterName)
+	}
+
+	resourcesToWatch := []watch.WatchedObject{}
+	defer func() {
+		r.EnsureMultiplesResourcesAreWatched(requestNamespacedName, r.Log, resourcesToWatch...)
+		r.Log.Debugf("watched backup schedule and policy resources: %v\r\n", r.WatchedResources)
+	}()
+
+	bSchedule, err := r.ensureBackupSchedule(ctx, service, backupScheduleRef, clusterName, &resourcesToWatch)
+	if err != nil {
+		return err
+	}
+
+	bPolicy, err := r.ensureBackupPolicy(ctx, service, *bSchedule.Spec.PolicyRef.GetObject(bSchedule.Namespace), &resourcesToWatch)
+	if err != nil {
+		return err
+	}
+
+	return r.updateBackupScheduleAndPolicy(ctx, service, projectID, clusterName, bSchedule, bPolicy)
+}
+
+func (r *AtlasDeploymentReconciler) ensureBackupSchedule(
+	ctx context.Context,
+	service *workflow.Context,
+	backupScheduleRef types.NamespacedName,
+	clusterName string,
+	resourcesToWatch *[]watch.WatchedObject,
+) (*mdbv1.AtlasBackupSchedule, error) {
+	bSchedule := &mdbv1.AtlasBackupSchedule{}
+	err := r.Client.Get(ctx, backupScheduleRef, bSchedule)
+	if err != nil {
+		return nil, fmt.Errorf("%v backupschedule resource is not found. e: %w", backupScheduleRef, err)
+	}
+
+	resourceVersionIsValid := customresource.ValidateResourceVersion(service, bSchedule, r.Log)
+	if !resourceVersionIsValid.IsOk() {
+		errText := fmt.Sprintf("backup schedule validation result: %v", resourceVersionIsValid)
+		r.Log.Debug(errText)
+		return nil, errors.New(errText)
+	}
+
+	bSchedule.UpdateStatus([]status.Condition{}, status.AtlasBackupScheduleSetDeploymentID(clusterName))
+
+	if err = r.Client.Status().Update(ctx, bSchedule); err != nil {
+		r.Log.Errorw("Failed to update BackupSchedule status", "error", err)
```

Comment:
Please do not start error messages with capital letters. See: https://github.com/golang/go/wiki/CodeReviewComments#error-strings